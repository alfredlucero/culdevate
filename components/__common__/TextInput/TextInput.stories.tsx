import React from "react";
import { storiesOf } from "@storybook/react";
import StoryFrame from "../StoryFrame";
import TextInput from "./index";

storiesOf("TextInput", module).add("Default", () => (
  <StoryFrame>
    <TextInput value="someTextInput" name="textInputName" valid={true} touched={true} type="text" onChange={() => {}} />
  </StoryFrame>
));
