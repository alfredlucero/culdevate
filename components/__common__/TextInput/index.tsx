import React from "react";
import styled from "styled-components";
import { culdevateDefaultTheme } from "../../defaultTheme";

const StyledTextInputContainer = styled.div`
  display: flex;
  flex-direction: column;
`;
StyledTextInputContainer.defaultProps = {
  theme: culdevateDefaultTheme,
};

interface StyledTextInputProps {
  isError: boolean;
}
const StyledTextInput = styled.input<StyledTextInputProps>`
  width: 100%;
  color: ${props => props.theme.colors.gray};
  padding: 1.5rem;
  border: 2px solid ${props => (props.isError ? props.theme.colors.red : props.theme.colors.lightGray)};
  border-radius: 5px;
`;
StyledTextInput.defaultProps = {
  theme: culdevateDefaultTheme,
};

export type TextInputType = "text" | "password" | "email";

export interface TextInputProps {
  value: string;
  id: string;
  name: string;
  valid: boolean;
  touched: boolean;
  type: TextInputType;
  onChange: (event: React.FormEvent<HTMLInputElement>) => void;
  label?: string;
  disabled?: boolean;
  infoMessage?: string;
  errorMessage?: string;
  placeholder?: string;
  required?: boolean;
}

const TextInput: React.FC<TextInputProps> = ({
  value,
  id,
  name,
  valid,
  touched,
  type,
  onChange,
  label = "",
  disabled = false,
  infoMessage = "",
  errorMessage = "",
  placeholder = "",
  required = true,
  ...passThroughProps
}) => {
  const infoId = infoMessage && `${id}-text-input-info`;
  const errorId = errorMessage && `${id}-text-input-error`;
  const isError = touched && !valid;

  return (
    <StyledTextInputContainer>
      <label htmlFor={id}>
        {label} {label && required && "(Required)"}
      </label>
      <StyledTextInput
        isError={isError}
        type={type}
        value={value}
        onChange={onChange}
        id={id}
        name={name}
        disabled={disabled}
        placeholder={placeholder}
        aria-describedby={isError ? errorId : infoId}
        {...passThroughProps}
      />
      {!isError && (
        <span data-testid="textInputInfo" id={infoId}>
          {infoMessage}
        </span>
      )}
      {isError && (
        <span data-testid="textInputError" id={errorId}>
          {errorMessage}
        </span>
      )}
    </StyledTextInputContainer>
  );
};

export default TextInput;
