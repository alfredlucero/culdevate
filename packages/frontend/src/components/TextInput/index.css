.text-input {
  @apply shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight;
}

.text-input:focus {
  @apply outline-none shadow-outline;
}

.text-input-error {
  @apply border-red-500;
}

.text-input-disabled {
  @apply bg-gray-200 opacity-75;
}

.text-input-label {
  @apply block text-gray-700 text-sm font-bold mb-2;
}

.text-input-info {
  @apply text-gray-700 text-xs italic mt-2;
}

.text-input-error-info {
  @apply text-red-500 text-xs italic mt-2;
}
