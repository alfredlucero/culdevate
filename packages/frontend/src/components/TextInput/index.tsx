import React from "react";
import cn from "classnames";
import { CommonProps } from "../commonProps";

interface TextInputProps extends CommonProps {
  onChange: (e: React.FormEvent<HTMLInputElement>) => void;
  onBlur?: (e: React.FocusEvent) => void;
  value: string;
  id: string;
  type: "text" | "password";
  required?: boolean;
  valid?: boolean;
  errorInfo?: string;
  textInfo?: string;
  label?: string;
  disabled?: boolean;
  placeholder?: string;
}

const textInputBaseClasses = [
  "shadow",
  "appearance-none",
  "border",
  "w-full",
  "py-2",
  "px-3",
  "text-gray-700",
  "leading-tight",
  "focus:outline-none",
  "focus:shadow-outline",
  "rounded",
];
const textInputErrorClasses = ["border-red-500"];
const textInputDisabledClasses = ["bg-gray-200", "opacity-75"];

const TextInput: React.FC<TextInputProps> = ({
  onChange,
  value,
  id,
  type,
  onBlur = () => {},
  required = false,
  valid = true,
  errorInfo = "",
  textInfo = "",
  label = "",
  disabled = false,
  placeholder = "",
  className = "",
  testId = "",
  ...passThroughProps
}) => {
  return (
    <div {...(className !== "" ? { className } : {})}>
      {label && (
        <label className="block text-gray-700 text-sm font-bold mb-2" htmlFor={id}>
          {label} {required ? "" : "(optional)"}
        </label>
      )}

      <input
        className={cn(
          textInputBaseClasses,
          !valid ? textInputErrorClasses : [],
          disabled ? textInputDisabledClasses : [],
        )}
        value={value}
        id={id}
        name={id}
        onChange={onChange}
        onBlur={onBlur}
        type={type}
        placeholder={placeholder}
        disabled={disabled}
        aria-required={required}
        {...(testId !== "" ? { "data-testid": testId } : {})}
        {...passThroughProps}
      />

      {valid && textInfo !== "" && <p className="text-gray-700 text-xs italic mt-2">{textInfo}</p>}

      {!valid && errorInfo !== "" && <p className="text-red-500 text-xs italic mt-2">{errorInfo}</p>}
    </div>
  );
};

export default TextInput;
