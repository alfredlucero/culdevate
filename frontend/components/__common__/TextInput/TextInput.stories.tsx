import React from "react";
import { storiesOf } from "@storybook/react";
import StoryFrame from "../StoryFrame";
import TextInput, { TextInputType } from "./index";

const requiredTextInputProps = {
  value: "someTextInput",
  id: "some-text-input-id",
  name: "textInputName",
  valid: true,
  touched: true,
  type: "text" as TextInputType,
  onChange: () => {},
};

storiesOf("common/TextInput", module)
  .add("Default", () => (
    <StoryFrame>
      <TextInput {...requiredTextInputProps} />
      <TextInput {...requiredTextInputProps} value="some.email@domain.com" type="email" />
      <TextInput {...requiredTextInputProps} type="password" />
    </StoryFrame>
  ))
  .add("Valid + Info Message", () => (
    <StoryFrame>
      <TextInput {...requiredTextInputProps} infoMessage="Some info message" />
    </StoryFrame>
  ))
  .add("Error + Error Message", () => (
    <StoryFrame>
      <TextInput {...requiredTextInputProps} valid={false} errorMessage="Some error message" />
    </StoryFrame>
  ))
  .add("Label + placeholder + disabled + required", () => (
    <StoryFrame>
      <TextInput
        {...requiredTextInputProps}
        label="Some label"
        placeholder="Some placeholder text"
        value=""
        disabled={true}
        required={true}
      />
    </StoryFrame>
  ));
