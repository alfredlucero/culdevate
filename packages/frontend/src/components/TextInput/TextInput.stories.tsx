import React from "react";
import { storiesOf } from "@storybook/react";
import TextInput from "./index";

storiesOf("Common/Text Input", module)
  .add("Default Text", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput id="text-input" value="Text Value" onChange={() => {}} type="text" />
    </div>
  ))
  .add("Default Password", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput id="password-input" value="Password Value" onChange={() => {}} type="password" />
    </div>
  ))
  .add("Required Label", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="Required Username"
        onChange={() => {}}
        type="text"
        label="Username"
        required={true}
      />
    </div>
  ))
  .add("Disabled Placeholder", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value=""
        placeholder="Enter a username..."
        onChange={() => {}}
        type="text"
        label="Username"
        disabled={true}
      />
    </div>
  ))
  .add("Inline Text Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="testuser123"
        onChange={() => {}}
        type="text"
        label="Username"
        valid={true}
        textInfo="Please provide a username."
        disabled={true}
      />
    </div>
  ))
  .add("Inline Text Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="testuser123"
        onChange={() => {}}
        type="text"
        label="Username"
        valid={true}
        textInfo="Please provide a username."
      />
    </div>
  ))
  .add("Inline Error Info", () => (
    <div style={{ width: "400px", padding: "30px" }}>
      <TextInput
        id="text-input"
        value="testuser123"
        onChange={() => {}}
        type="text"
        label="Username"
        valid={false}
        errorInfo="Please provide a username."
      />
    </div>
  ));
