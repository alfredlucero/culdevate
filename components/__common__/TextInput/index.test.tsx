import React from "react";
import { render, cleanup } from "react-testing-library";
import TextInput, { TextInputType } from "./index";

afterEach(cleanup);

describe("<TextInput />", () => {
  const requiredTextInputProps = {
    value: "someTextInput",
    name: "textInputName",
    valid: true,
    touched: true,
    type: "text" as TextInputType,
    onChange: () => {},
  };

  test("should render the text input", () => {
    const { container } = render(<TextInput {...requiredTextInputProps} />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
