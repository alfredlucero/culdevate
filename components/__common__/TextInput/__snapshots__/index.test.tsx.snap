// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextInput /> should render the text input 1`] = `
<div
  class="sc-bdVaJa bPGLpq"
>
  <label
    for="some-text-input-id"
  >
    
     
    
  </label>
  <input
    aria-describedby=""
    class="sc-bwzfXH gAQYSD"
    id="some-text-input-id"
    name="textInputName"
    placeholder=""
    type="text"
    value="someTextInput"
  />
  <em
    class="sc-htpNat kbZKBX"
    data-testid="textInputInfo"
    id=""
  />
</div>
`;
